src/main-process/file-recovery-service.js

```diff
@@ -0,0 +1,119 @@
+    this.recoveryDirectory = recoveryDirectory
+    this.recoveryFilesByFilePath = new Map()
+    this.recoveryFilesByWindow = new WeakMap()
+  }
+
+  willSavePath (window, path) {
+    if (!fs.existsSync(path)) return
+
+    let recoveryFile = this.recoveryFilesByFilePath.get(path)
+    if (recoveryFile == null) {
+      recoveryFile = new RecoveryFile(
+        path,
+        Path.join(this.recoveryDirectory, RecoveryFile.fileNameForPath(path))
+      )
+      this.recoveryFilesByFilePath.set(path, recoveryFile)
+    }
+
+    try {
+      recoveryFile.retain()
+    } catch (err) {
+      console.log(`Couldn't retain ${recoveryFile.recoveryPath}. Code: ${err.code}. Message: ${err.message}`)
+    }
+
+    if (!this.recoveryFilesByWindow.has(window)) {
+      this.recoveryFilesByWindow.set(window, new Set())
+    }
+    this.recoveryFilesByWindow.get(window).add(recoveryFile)
+  }
+
+  didSavePath (window, path) {
+    const recoveryFile = this.recoveryFilesByFilePath.get(path)
+    if (recoveryFile != null) {
+      try {
+        recoveryFile.release()
+      } catch (err) {
+        console.log(`Couldn't release ${recoveryFile.recoveryPath}. Code: ${err.code}. Message: ${err.message}`)
+      }
+      if (recoveryFile.isReleased()) this.recoveryFilesByFilePath.delete(path)
+      this.recoveryFilesByWindow.get(window).delete(recoveryFile)
+    }
+  }
+
+  didCrashWindow (window) {
+    if (!this.recoveryFilesByWindow.has(window)) return
+
+    for (const recoveryFile of this.recoveryFilesByWindow.get(window)) {
+      try {
+        recoveryFile.recoverSync()
+      } catch (error) {
+        const message = 'A file that Atom was saving could be corrupted'
+        const detail =
+          `There was a crash while saving "${recoveryFile.originalPath}", so this file might be blank or corrupted.\n` +
+          `Atom couldn't recover it automatically, but a recovery file has been saved at: "${recoveryFile.recoveryPath}".`
+        console.log(detail)
+        dialog.showMessageBox(window.browserWindow, {type: 'info', buttons: ['OK'], message, detail})
+      } finally {
+        this.recoveryFilesByFilePath.delete(recoveryFile.originalPath)
+      }
+    }
+
+    this.recoveryFilesByWindow.delete(window)
+  }
+
+  didCloseWindow (window) {
+    this.recoveryFilesByWindow.delete(window)
+  }
+}
+
+class RecoveryFile {
+  static fileNameForPath (path) {
+    const extension = Path.extname(path)
+    const basename = Path.basename(path, extension).substring(0, 34)
+    const randomSuffix = crypto.randomBytes(3).toString('hex')
+    return `${basename}-${randomSuffix}${extension}`
+  }
+
+  constructor (originalPath, recoveryPath) {
+    this.originalPath = originalPath
+    this.recoveryPath = recoveryPath
+    this.refCount = 0
+  }
+
+  storeSync () {
+    fs.writeFileSync(this.recoveryPath, fs.readFileSync(this.originalPath))
+  }
+
+  recoverSync () {
+    fs.writeFileSync(this.originalPath, fs.readFileSync(this.recoveryPath))
```

Comment:
If the files are large, will `storeSync` and `recoverSync` greatly increase memory usage? I wonder if reading some number of bytes at a time, or using some synchronous version of `pipe`, would be better?